### PR TITLE
[android] add projects to record AOT profiles

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Context: https://github.com/jonathanpeppers/Mono.Profiler.Android#usage-of-the-aot-profiler

Adds projects and instructions for creating AOT profiles for Android.
Additionally, add a new profile for `maui-blazor` to be used when
`$(UsingMicrosoftNETSdkRazor)` is `true`.

`CommonMethods.cs` has various APIs that we want to include in the
built-in AOT profile. Right now, it most importantly includes usage of
`HttpClient` and `System.Threading.Tasks`.

`maui.aotprofile.txt` is an alphabetical list of methods in the
profile. We can track changes to this file over time.

### Results ###

`dotnet new maui` on a Pixel 5:

    02-23 09:48:40.822  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +470ms
    02-23 09:48:42.056  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +485ms
    02-23 09:48:43.262  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +465ms
    02-23 09:48:44.438  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +471ms
    02-23 09:48:45.667  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +473ms
    02-23 09:48:46.909  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +458ms
    02-23 09:48:48.100  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +462ms
    02-23 09:48:49.312  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +463ms
    02-23 09:48:50.538  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +461ms
    02-23 09:48:51.719  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +465ms
    Average(ms): 467.3
    Std Err(ms): 2.47229088543849
    Std Dev(ms): 7.81807023646003

Same project using MAUI Preview 13 was `Average(ms): 490.9`.

`dotnet new maui-blazor`:

    02-23 09:51:23.245  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +570ms
    02-23 09:51:24.480  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +595ms
    02-23 09:51:25.656  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +574ms
    02-23 09:51:26.870  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +577ms
    02-23 09:51:28.078  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +584ms
    02-23 09:51:29.243  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +556ms
    02-23 09:51:30.457  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +570ms
    02-23 09:51:31.699  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +596ms
    02-23 09:51:32.896  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +578ms
    02-23 09:51:34.117  1802  1962 I ActivityTaskManager: Displayed com.companyname.foo/crc64808a40cc7e533249.MainActivity: +578ms
    Average(ms): 577.8
    Std Err(ms): 3.76769897358528
    Std Dev(ms): 11.9145102944081

Same project using MAUI Preview 13 was `Average(ms): 664.0`.

### PR Checklist ###

- [ ] Targets the correct branch 
- [ ] Tests are passing (or failures are unrelated)
- [ ] Targets a single property for a single control (or intertwined few properties)
- [ ] Adds the property to the appropriate interface
- [ ] Avoids any changes not essential to the handler property
- [ ] Adds the mapping to the PropertyMapper in the handler
- [ ] Adds the mapping method to the WinUI, Android, iOS, and Standard aspects of the handler
- [ ] Implements the actual property updates (usually in extension methods in the Platform section of Core)
- [ ] Tags ported renderer methods with [PortHandler]
- [ ] Adds an example of the property to the sample project (MainPage)
- [ ] Adds the property to the stub class
- [ ] Implements basic property tests in DeviceTests

#### Does this PR touch anything that might affect accessibility?

No

### CODE REVIEW

File: src/ProfiledAot/maui.aotprofile.txt [hunk 1]
```diff
@@ -0,0 +1,5586 @@
+Methods:
+	Android.App.Activity Microsoft.Maui.Handlers.ElementHandler`2<Microsoft.Maui.IWindow, Android.App.Activity>:get_PlatformView ()
+	Android.App.Activity Microsoft.Maui.Handlers.WindowHandler:CreatePlatformElement ()
+	Android.App.Activity Microsoft.Maui.Platform.ContextExtensions:GetActivity (Android.Content.Context)
```
Comment (jonathanpeppers) [inline src/ProfiledAot/maui.aotprofile.txt]: This is a text version of the above `maui.aotprofile`.

File: src/ProfiledAot/README.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
+# Profiled AOT support for Android
```
Comment (jonathanpeppers) [inline src/ProfiledAot/README.md:1]: This file has the instructions on how to update these files, basically:

    $ ./bin/dotnet/dotnet build -t:Record -r android-arm64 src/ProfiledAot/ProfiledAotMaui/ProfiledAotMaui.csproj
    $ ./bin/dotnet/dotnet build -t:Record -r android-arm64 src/ProfiledAot/ProfiledAotMauiBlazor/ProfiledAotMauiBlazor.csproj

**Comment by jonathanpeppers** [comment]
In-tree tests are failing... because I moved the `.aotprofile` files to `src\ProfiledAot`. Will look into it.

File: src/ProfiledAot/CommonMethods.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+// This class represents common code paths we always want to AOT
+
+// System.Threading.Tasks.Task
+// System.Net.Http.HttpClient
+// System.Console
```
Comment (eerhardt) [inline src/ProfiledAot/CommonMethods.cs:5]: What calls System.Console in a default app?

Comment (jonathanpeppers) [reply src/ProfiledAot/CommonMethods.cs:5]: I could maybe remove this one.

File: src/ProfiledAot/ProfiledAotMaui/App.xaml [hunk 4]
```diff
@@ -0,0 +1,26 @@
+﻿<?xml version = "1.0" encoding = "UTF-8" ?>
+<Application xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
```
Comment (jonathanpeppers) [inline src/ProfiledAot/ProfiledAotMaui/App.xaml:2]: I'm going to try to rework these projects so it runs `dotnet new` into a folder, and the `Directory.Build.targets` sets everything up.

File: src/ProfiledAot/CommonMethods.cs (hunk 3 above)
Comment (eerhardt) [reply src/ProfiledAot/CommonMethods.cs:5]: I'm not necessarily asking for it to be removed. More for education on "why would something in the default app call System.Console?" If we are, what is the scenario? etc.

Comment (jonathanpeppers) [reply src/ProfiledAot/CommonMethods.cs:5]: I saw it in the list when looking at JIT'd methods of the podcast app, but they might just have a random call somewhere.

File: src/ProfiledAot/build.proj [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    <Intermediate>src/obj/$(App)/</Intermediate>
+    <CSProj>$(Intermediate)$(App).csproj</CSProj>
+    <EnvVars>DOTNET_MULTILEVEL_LOOKUP=0</EnvVars>
+  </PropertyGroup>
+  <Target Name="Record">
+    <Error Condition=" '$(App)' != 'maui' and '$(App)' != 'maui-blazor' " Text="Please pass -p:App=maui or -p:App=maui-blazor" />
+    <RemoveDir Directories="$(Intermediate)" />
+    <Exec Command="&quot;$(DotNetToolPath)&quot; new $(App) -o $(Intermediate)" EnvironmentVariables="$(EnvVars)" />
+    <!-- MainPage.xaml.cs uses CommonMethod.cs -->
+    <Copy SourceFiles="src/MainPage-$(App).cs" DestinationFiles="$(Intermediate)MainPage.xaml.cs" />
```
Comment (jonathanpeppers) [inline src/ProfiledAot/build.proj:14]: This now invokes the template and copies one file over top. So a lot less code now.

File: src/ProfiledAot/README.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+## Updating Profiles
+
+Build MAUI following the instructions at [DEVELOPMENT.md][0]. Make
+sure to build with `--configuration=Release`.
+
+Run the `Record` target on each project:
+
+    $ ./bin/dotnet/dotnet build src/ProfiledAot/build.proj -p:App=maui
+    $ ./bin/dotnet/dotnet build src/ProfiledAot/build.proj -p:App=maui-blazor
```
Comment (jonathanpeppers) [inline src/ProfiledAot/README.md:15]: Instructions have changed now as well.

File: .nuspec/Microsoft.Maui.Controls.targets [hunk 7]
```diff
@@ -25,6 +25,7 @@
 
 	<ItemGroup Condition="'$(AndroidEnableProfiledAot)' == 'true' and '$(MauiUseDefaultAotProfile)' != 'false'">
 		<AndroidAotProfile Include="$(MSBuildThisFileDirectory)maui.aotprofile" />
+		<AndroidAotProfile Include="$(MSBuildThisFileDirectory)maui-blazor.aotprofile" Condition=" '$(UsingMicrosoftNETSdkRazor)' == 'true' " />
```
Comment (eerhardt) [inline .nuspec/Microsoft.Maui.Controls.targets:28]: Does it hurt to just always use both profiles? Would there be any advantage to doing that?

Comment (jonathanpeppers) [reply .nuspec/Microsoft.Maui.Controls.targets:28]: It seems like that wouldn't hurt, I see some common stuff like `TaskCompletionSource` and `System.Text.Json` in the `maui-blazor` profile.

Comment (eerhardt) [reply .nuspec/Microsoft.Maui.Controls.targets]: I can easily see non-Blazor apps using that code.

File: src/ProfiledAot/README.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+To verify what methods are AOT'd, clear the log and enable AOT logging:
+
+    $ adb logcat -c
+    $ adb shell setprop debug.mono.log default,timing=bare,assembly,mono_log_level=debug,mono_log_mask=aot
+
+Restart the app, and you should be able to see messages like:
+
+    $ adb logcat -d | grep AOT
+    02-23 09:03:46.327 10401 10401 D Mono    : AOT: FOUND method Microsoft.AspNetCore.Components.WebView.Maui.BlazorWebView:.ctor () [0x6f9efd0150 - 
+0x6f9efd0340 0x6f9efd260c]
```
Comment (eerhardt) [inline src/ProfiledAot/README.md:59]: (nit) this is not formatted correctly when viewing the markdown doc.

Sidebar: Is there a reason you aren't using triple-ticks to encapsulate these in a code block?

````

```
$ adb logcat -d | grep AOT
02-23 09:03:46.327 10401 10401 D Mono    : AOT: FOUND method Microsoft.AspNetCore.Components.WebView.Maui.BlazorWebView:.ctor () [0x6f9efd0150 - 
0x6f9efd0340 0x6f9efd260c]
```

````

Same problem on line 68: `(intptr,intptr,intptr).`

**Review by eerhardt — APPROVED** [review]
Looks great! Just one minor formatting issue in the README.

File: src/ProfiledAot/README.md (hunk 8 above)
Comment (jonathanpeppers) [reply src/ProfiledAot/README.md]: I think my terminal wrapped some lines when I copy/pasted these. Updated to use fenced code blocks, too.

**Pull request merged**